Reworks sabotage progression traitor objective

Jacquerel opened the pull request:
## About The Pull Request

This one is a little bit of a more abstract change than my last traitor objective rebalancing PR.

The idea here is that I think that the objectives which have you sabotage a frequently used machine (lathes or the ORM, but we can maybe expand it to others following this?) are potentially pretty good but limitations in how the objective are tracked mean that people go about them in a boring way which rarely inconveniences anyone.

I have reworked these objectives so that you can choose to apply a provided item to the machine over 10 seconds. The next interaction following this (clicking on it, or in the case of the ORM loading any ore into it) will cause the machine to explode, destroying itself and lightly damaging people nearby (explosion damage is... variable, but it's a light explosion so it won't instantly kill anyone).
There is a 0.5 second grace period between the bomb triggering and the explosion actually happening, which players can use to perform some kind of tactical reaction or simply to feel regret.

Here's a video:
https://www.youtube.com/watch?v=xNIaHsA9YFE

Trapped machines will intermittently beep, though at a volume which is hard to detect from more than two tiles away, and can be defused by using a screwdriver on the machine.
If you choose to do this then you will get a little bonus to TC and reputation, but only if the trap is successfully sprung. This means that sabotaging a machine in an empty department is unlikely to lead to quick success.
You can also still do it with screwdriver and crowbar way if you are boring but this means you will only get reputation and not any telecrystals.

The two objectives to destroy a radio server or the research machine remain as they were.

Also I refactored how we set which machines are targets for these objectives so it doesn't iterate over every machine in the world every time.

## Why It's Good For The Game

Traitor objectives (and antagonist objectives in general) should:
- Present you with a somewhat engaging task
- Encourage you to interact with at least one other player (at least by "attempting to sneak past them")
- Move the state of the round forward in some way, usually by creating inconvenience for at least one other person

Prior to this change you could, and people often did, complete this objective by accessing a machine that you already _have_ access to, using a screwdriver and crowbar on it, and then reassembling it.
This prompts interaction with nobody, and inconveniences nobody.

Following this change people will be encouraged to:
- Actually destroy the machine.
- Probably harm somebody.
- Potentially cause some collateral damage.
- Make loud sounds and flashing lights to delight people of all ages.

Even if you opt to set the device in a little-used department and then blow yourself up, that will prompt you to go and interact with someone in medbay and will also look pretty suspicious in a way that's much harder to track with the previous objective.

Also I might reuse this component again in the future, you never know.

## Changelog

:cl: Jacquerel, sprite by J(Clearly Lying)
balance: Traitor objectives which ask you to destroy lathes or the ORM provide an optional bonus if you _booby trap_ the machine using a provided tool. This causes the machine to explode the next time it is interacted with.
balance: Performing the objective without rigging the machine to explode awards no TC, using it awards more TC than it did previously.
/:cl:

Jacquerel committed:
Split up files

Jacquerel committed:
Start adding my thing

Jacquerel committed:
Mostly finished the item

Jacquerel committed:
Item is all done

Jacquerel committed:
Componentpilled

Jacquerel committed:
Quick addition to make sure this actually sabotages something

castawaynont commented:
I feel like nobody will ever set it off if it makes the beeps.

ATH1909 commented:
a light explosion feels... pretty pathetic? that's gonna do, what, like, 30 damage?

if it's something that someone has to interact with to trigger, has this many tells, and takes 10 seconds to set up, I think you could bump it up to a heavy explosion (120 damage or so).

Ryusenshu commented:
- Does it break the board? (since only one replacement lathe board in the lockers)
- Do silicons trigger the lathe bomb?

In any case, not much of fan, but i never was of that objective

> a light explosion feels... pretty pathetic? that's gonna do, what, like, 30 damage?
> 
> if it's something that someone has to interact with to trigger, has this many tells, and takes 10 seconds to set up, I think you could bump it up to a heavy explosion (120 damage or so).

Bigger Explosion would make it way too good for some thing you GAIN TC from

flowercuco commented:
> * Does it break the board? (since only one replacement lathe board in the lockers)

it has the potential to break the board, but will not always due to how the bomb works iirc

Vect0r2 commented:
What if more than one person has a protolathe mission?

Watermelon914 commented on code/datums/components/machine_booby_trap.dm:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Attached to an atom, creates an explosion when it is interacted with
+ */
+/datum/component/
```
you haven't given this component a typepath

Jacquerel replied:
How the hell did I do that

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,133 @@
+		/datum/traitor_objective/sabotage_machinery = 1,
+	)
+
+/datum/traitor_objective/sabotage_machinery
```
For this objective, I think it would be interesting if you let them destroy it normally but at a reduction in rewards (both in TC and progression)
I prefer traitor objectives to be more open-ended.

Jacquerel replied:
My problem with this is that there's no method I can find of allowing open-ended destruction which doesn't simply include clicking on the machine once with a screwdriver and then a crowbar, which causes the machine to cease to exist as far as code is concerned.
If we'd prefer to leave it open-ended then I completely understand the reasoning but it'd probably be best to just close this PR as a failed experiment, as preventing that behaviour was sort of the sole purpose.

Jacquerel replied:
I guess the alternative is to make the objective explicit about destroying the machine board rather than the machine?
As other people have pointed out, there are currently only two machine boards for lathes in existence on any map and no means of procuring more.
This is either a positive (you should simply protect it better) or a negative (well now I can't print any department-specific gear), although I could also simply add a means of acquiring more in the same PR.

Watermelon914 replied:
That's fine though, clicking with screwdriver and then crowbar is fine. Just make it give reduced TC and progression, players can choose the lazy way out or they can make extra TC/progression, it's up to them.

Jacquerel replied:
My sort of goal here is to reduce or remove instances where completing an objective has literally no impact at all on anyone.
When you deconstruct a machine with tools you can simply immediately reconstruct it, meaning that you did not inconvenience anyone at all and as essentially no crime happened nobody has any reason to investigate either, unless you are caught in the act (a span of about 15 seconds) nobody will ever know.

I can add it back in with reduced TC and then also try to make it so that non-roundstart protolathes need to be manually linked to the vault in the same manner as the ORM so that deconstructing it still leaves some external task to complete before the machine is functional?
That would address both of our issues, but might constitute a different PR.

Vect0r2 replied:
It does *do* something, which is unlink from the ore silo.

Jacquerel replied:
> It does _do_ something, which is unlink from the ore silo.

You know for some reason I was CONVINCED that only the ORM did this but you're right it does apply to all three. 
In that case yeah I'll just put the signal back as a reduced reward.

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,133 @@
+		possible_jobs -= target_head
+
+		chosen_job = target_head
+		for(var/obj/machinery/machine as anything in GLOB.machines)
+			if(istype(machine, machine_to_find) && is_station_level(machine.z))
+				possible_machines += machine
```
You shouldn't be generating the objectives like this, this is laggy.
Generate a list of possible targets like how the steal objective works and access the types from this list.

Watermelon914 replied:
nvm you can ignore this

Jacquerel replied:
Too late I agree so I have written it

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,133 @@
+
+
+/// Item which you use on a machine to cause it to explode next time someone interacts with it
+/obj/item/traitor_machine_trapper/
```
```suggestion
/obj/item/traitor_machine_trapper
```

Watermelon914 commented on code/modules/mining/machine_redemption.dm:
```diff
@@ -156,6 +156,8 @@
 	else
 		return
 
+	SEND_SIGNAL(src, COMSIG_ORM_COLLECTED_ORE)
```
for this signal, it would make sense to pass in the ore as one of the arguments, even if you don't end up using it.

Jacquerel commented:
> What if more than one person has a protolathe mission?

One person succesfully completing the objective will mark it as complete for everyone who currently has the objective, and make it invalid to generate in the future. This means that it _shouldn't_ lead to both boards being destroyed.

Jacquerel commented:
> I feel like nobody will ever set it off if it makes the beeps.

It's a reasonable concern, I think they are both quiet and intermittent enough that they're pretty easy to miss and I didn't want it to be _completely_ undetecable but if it turns out that this just results in them never going off it's easy enough to remove.

Jacquerel committed:
Don't add my new properties to literally every component in the codebase
Pass gathered ore into new signal
Put sabotage objectives back into the same file
Streamline target generation

ATH1909 commented:
> Bigger Explosion would make it way too good for some thing you GAIN TC from
> 
> Im not a fan of adding a free bomb Like, we had all the talk back then about "early bombings are bad" and now we are here again

Not a BIGGER explosion, just a more intense one. Explosion intensity radii are independent of each other- you can, for example, code an explosion with a strong explosion radius of 2 but a light explosion radius of only 3. this wouldn't make any hull breaches (due to having a devastating explosion radius of 0), to my knowledge, and would only significantly damage things in a 3x3 square (with light damage extending to a 5x5 square). I don't think that's anywhere close to the same scale as, say, a syndicate bomb. hell, I think our syndicate minibombs might be more destructive than that.

SomeoneYouProbablyKnow commented:
You wanna use the bug sprite for the devise? i could probably whip up a bad sprite specifically for this objective

Jacquerel commented:
> You wanna use the bug sprite for the devise? i could probably whip up a bad sprite specifically for this objective

I think it kind of fits and it's being removed from its use to bug rooms soon anyway, but if you have a strong idea be my guest although I can't guarantee I'll use it if I like the bug sprite more 😅 
I'd have made one myself but I couldn't think of something which wasn't pretty similar to what I already have, and it does the job.

Jacquerel committed:
New tool now gives a bonus rather than being the only way to complete it

Jacquerel commented:
As discussed, this is now an optional extra rather than the only way to complete those objectives.
This makes the work a little less justified but it still comes with some code improvements so I don't think it was a waste of time, and I do think using the tool is often more fun.

Jacquerel committed:
Adds a sprite

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,235 @@
+
+/datum/traitor_objective/sabotage_machinery/trap
+	name = "Sabotage the %MACHINE%"
+	description = "Destroy the %MACHINE% to cause disarray and disrupt the operations of the %JOB%'s department. If you can have another crew member destroy the machine using the provided booby trap, you will be rewarded with an additional %PROGRESSION% reputation and %TC% telecrystal."
```
```suggestion
	description = "Destroy the %MACHINE% to cause disarray and disrupt the operations of the %JOB%'s department. If you get another crew member to destroy the machine using the provided booby trap, you will be rewarded with an additional %PROGRESSION% reputation and %TC% telecrystal."
```

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,235 @@
+		JOB_SHAFT_MINER = /obj/machinery/mineral/ore_redemption,
+	)
+
+	/// Bonus reward to grant if you booby trap successfully
+	var/bonus_tc = 1
+	/// Bonus reputation to grant if you booby trap successfully
+	var/bonus_rep = 2 MINUTES
```
I don't mind beefing up the bonus tc, to 2-3 and then reducing the actual awarded TC

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,235 @@
+		JOB_SCIENTIST = /obj/machinery/rnd/server,
+	)
+
+/datum/traitor_objective/sabotage_machinery/trap
```
You should explicitly set the telecrystal reward and progression for each non-abstract objective so that it's clear how much they award. Abstract types should not have a telecrystal reward and progression set so that the unit test fails when an objective doesn't explicitly set their own telecrystal reward and progression.

Jacquerel replied:
I'll be honest I was kind of thinking that when I did it, so this makes sense to me.

Jacquerel committed:
Bonus objective is now the sole source of TC
Explicitly set rewards and requirements on non-abstract types
Minor desc changes

JohnFulpWillard commented on code/datums/components/machine_booby_trap.dm:
```diff
@@ -0,0 +1,102 @@
+	datum/callback/on_triggered_callback = null,
+	datum/callback/on_defused_callback = null,
+)
+	if(!isatom(parent))
```
shouldn't it be ``ismachine`` if it's gonna be named machine booby trap

Jacquerel committed:
Renames component

Vect0r2 commented:
What happens if somebody disables the bomb?

Jacquerel commented:
> What happens if somebody disables the bomb?

It falls out and then either you beat them up and put it back in or you destroy the machine in some other fashion which doesn't earn you extra TC.

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,235 @@
+	)
+
+	/// Bonus reward to grant if you booby trap successfully
+	var/bonus_tc = 2
+	/// Bonus reputation to grant if you booby trap successfully
+	var/bonus_rep = 2 MINUTES
```
I think these should be variable. Honestly, the bonus rep is kinda disappointing.
Can you instead set the progression_reward to something very low and beef up the bonus_rep?
Also don't call it rep, call it progression in code. It's important to remember that reputation is the visual representation of progression. Progression is the time representation of reputation.

Watermelon914 commented on code/game/machinery/_machinery.dm:
```diff
@@ -184,6 +184,9 @@
 	if(HAS_TRAIT(SSstation, STATION_TRAIT_BOTS_GLITCHED))
 		randomize_language_if_on_station()
 
+	if(mapload && !GLOB.objective_machine_handler.targets_confirmed)
+		add_as_sabotage_target()
```
You should move this to a global signal instead and handle it via datum.

Watermelon914 commented on code/modules/antagonists/traitor/objectives/sabotage_machinery.dm:
```diff
@@ -0,0 +1,235 @@
+	forceMove(get_turf(machine))
+	visible_message(span_warning("A [src] falls out from the [machine]!"))
+
+/// Datum which manages references to things we are instructed to destroy
+GLOBAL_DATUM_INIT(objective_machine_handler, /datum/objective_target_machine_handler, new())
```
Ideally, define your global variables at the top of a file

Jacquerel committed:
Reduce normal progression, increase bonus progression
Refactor machine marking to use signals

Jacquerel committed:
Don't add more machines after the round has started (traitors won't know where they are)
